fix(core): Support extending differs from root `NgModule`

mhevery opened the pull request:
Differs tries to inject parent differ in order to support extending.
This does not work in the 'root' injector as the provider overrides the
default injector. The fix is to just assume standard set of providers
and extend those instead.

PR Close #25015
Issue close #11309 `Can't extend IterableDiffers`
Issue close #18554 `IterableDiffers.extend is not AOT compatible` (This is fixed because we no longer have an arrow function in the  factory but a proper function which can be imported.)


## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

AndrewKushnir reviewed (approved):
LGTM 👍

AndrewKushnir reviewed (approved):
Reviewed-for: size-tracking

AndrewKushnir commented on packages/forms/src/directives/reactive_directives/form_control_directive.ts:
```diff
@@ -125,7 +125,9 @@ export class FormControlDirective extends NgControl implements OnChanges {
     }
     if (isPropertyUpdated(changes, this.viewModel)) {
-      _ngModelWarning('formControl', FormControlDirective, this, this._ngModelWarningConfig);
+      if (typeof ngDevMode === 'undefined' || ngDevMode) {
+        _ngModelWarning('formControl', FormControlDirective, this, this._ngModelWarningConfig);
+      }
```
@mhevery it looks like these changes are unrelated (from a different PR that has already landed)?

mhevery replied:
Sorry, fixed

AndrewKushnir commented on packages/forms/src/directives/reactive_directives/form_control_name.ts:
```diff
@@ -145,7 +145,9 @@ export class FormControlName extends NgControl implements OnChanges, OnDestroy {
     if (!this._added) this._setUpControl();
     if (isPropertyUpdated(changes, this.viewModel)) {
-      _ngModelWarning('formControlName', FormControlName, this, this._ngModelWarningConfig);
+      if (typeof ngDevMode === 'undefined' || ngDevMode) {
+        _ngModelWarning('formControlName', FormControlName, this, this._ngModelWarningConfig);
+      }
```
These changes look unrelated too.

mhevery replied:
sorry, fixed

AndrewKushnir reviewed (commented):
@mhevery just noticed that there is a couple unrelated changes now. Could you please have a look when you get a sec?

mhevery committed:
fix(core): Support extending differs from root `NgModule`

Differs tries to inject parent differ in order to support extending.
This does not work in the 'root' injector as the provider overrides the
default injector. The fix is to just assume standard set of providers
and extend those instead.

PR close #25015
Issue close #11309 `Can't extend IterableDiffers`
Issue close #18554 `IterableDiffers.extend is not AOT compatible`
  (This is fixed because we no longer have an arrow function in the
  factory but a proper function which can be imported.)

mhevery commented:
[presubmit](https://test.corp.google.com/ui#id=OCL:345783570:BASE:345783598:1607129410548:e1caa75a)

petebacondarwin reviewed (approved):
Reviewed-for: size-tracking
